### ISSUE TITLE
BanHTTP Tests do not work on various platforms

### ISSUE THREAD

**Issue by mehmetf** [dart-lang/sdk#41328]
My CL was reverted with a5611cdda2088ed6dc0aab9c9ae65d3cbe926537 because tests are timing out on Windows. I don't have a Windows development machine (yet) so I am not sure how to make them pass. This is a placeholder issue to discuss with @sortie and figure out what to do next.

My current theory is that:

```dart
  final server = await HttpServer.bind(Platform.localHostname, 0);
  final uri = Uri(scheme: 'http', host: Platform.localHostname, port: server.port);
  await asyncTest(() => httpClient.getUrl(uri));
```

does not work on Windows due to use of localHostname. If other HTTP tests are passing, it could be because they all use localhost or loopback addresses. I cannot use these to test this functionality since loopback addresses are always allowed.

Several ideas come to mind:
- Disallow loopback based on an environment flag. [Not preferred since it reduces coverage and increases complexity]
- Disable tests on windows. There's no platform specific code in that CL. 
- Figure out a way to allow binding to the host name rather than localhost. [Preferred]

**Comment by mehmetf** [comment]
The tests failed on AOT Android tests as well.

This is a bigger mystery since the test is failing with `SocketException` when I try to ban HTTPS via a zone variable:

```dart
  asyncExpectThrows(
      () async => await runZoned(() => httpClient.getUrl(uri),
          zoneValues: {#dart.library.io.allow_http: false}),
      (e) => e is StateError);
```

Same in the embedder test:

```dart
// Passes
  asyncExpectThrows(
      () async => await httpClient.getUrl(httpUri), (e) => e is StateError);
// Passes
  asyncExpectThrows(
      () async => await runZoned(() => httpClient.getUrl(httpUri),
          zoneValues: {#dart.library.io.allow_http: 'foo'}),
      (e) => e is StateError);
// Throws `OSError` ?
  asyncExpectThrows(
      () async => await runZoned(() => httpClient.getUrl(httpUri),
          zoneValues: {#dart.library.io.allow_http: false}),
      (e) => e is StateError);
```

**Comment by mehmetf** [comment]
Looking at test scripts, it seems like these tests are running on a phone. They also contain different exceptions. 

https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket.appspot.com/8884020471788653472/+/steps/test_results/0/logs/new_test_failures__logs_/0

https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket.appspot.com/8884022100584288160/+/steps/test_results/0/logs/new_test_failures__logs_/0

**Comment by mehmetf** [comment]
OK. I resolved the windows problem. Basically `httpClient.close` needs to be called explicitly on Windows. I also restructured the tests to be more readable.

Unfortunately, I am unable to solve the Android phone problem. I tried using `NetworkInterface` to find a local IP to bind to but the phone does not provide such an interface. So I don't think this can be tested on phone.

Is there a way to ban tests on certain platforms?
